app/models/user.rb

```diff
@@ -2,9 +2,70 @@ class User < ApplicationRecord
   has_many :posts, dependent: :destroy
   has_many :comments, dependent: :destroy
   has_many :likes, dependent: :destroy
+  has_many :friendships, foreign_key: "creator_id", dependent: :destroy
+  has_many :inverse_friendships, class_name: "Friendship", foreign_key: "recipient_id", dependent: :destroy
   belongs_to :location, required: false
+  validates :email, uniqueness: true, presence: true
+  validates :name, presence: true
   # Include default devise modules. Others available are:
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
   devise :database_authenticatable, :registerable,
-         :recoverable, :rememberable, :validatable
+         :recoverable, :rememberable, :validatable,
+         :omniauthable, omniauth_providers: %i[facebook]
+
+
+  def self.new_with_session(params, session)
+    super.tap do |user|
+      if data = session["devise.facebook_data"] && session["devise.facebook_data"]["extra"]["raw_info"]
+        user.email = data["email"] if user.email.blank?
+      end
+    end
+  end
+
+  def confirmed_friends
+    created_friendships = Array.new
+    created_friendships = friendships.select { |i| i.status == true}.map do |f|
+      f.recipient
+    end
+
+    received_friendships = Array.new
+    received_friendships = inverse_friendships.select { |i| i.status == true }.map do |f|
+    f.creator
+    end
+
+    (created_friendships + received_friendships)
+  end
+
+  def find_friendship(user)
+    friendship = self.friendships.find{|f| f.recipient == user} || self.inverse_friendships.find{|f| f.creator == user}
+  end
+
+  def confirmed_friends_ids
+    confirmed_friends.map(&:id)
+  end
+
+  def pending_friendships
```

Comment:
- [ ] **`[OPTIONAL]`** To further prove my point on the importance of using `has_many: through` .If we have used `has_many :friends, through: confirmed_friendships` this automatically creates `friend_ids` method that returns a relation with the ids of all friends.